opentelemetry-api/src/opentelemetry/_metrics/__init__.py

```diff
@@ -129,17 +130,30 @@ def version(self):
     def schema_url(self):
         return self._schema_url
 
-    # FIXME check that the instrument name has not been used already
+    def _check_instrument_id(self, name, type_, unit, description):
+
+        instrument_id = "".join(
+            [name.strip().lower(), str(type_), unit, description]
+        )
+
+        if instrument_id in self._instrument_ids:
+            _logger.warning(
+                "Instrument id %s has been used already", instrument_id
+            )
+        else:
+
+            with self._instrument_ids_lock:
+                self._instrument_ids.add(instrument_id)
 
     @abstractmethod
     def create_counter(self, name, unit="", description="") -> Counter:
-        pass
+        self._check_instrument_id(name, Counter, unit, description)
```

Comment:
> I would prefer to just leave this as `pass` and allow the subclasses to call `_check_instrument_id()` themselves.

We can do that if you prefer. I find `super` to be more intuitive because it tells the user that the code in the parent method is useful for the code in the child method.

> If we provide an implementation for an abstractmethod which returns a Counter instance, it should return a Counter instance.

We should not expect that from a parent method. There are not unimplemented methods in Python, even an parent method with only `pass` is being implemented and its implementation is returning `None`, not `Counter`. Also, typing in Python is secondary, the only reason why someone may consider the parent method to have to return `Counter` is because there is a typing hint in the method signature. We could remove this signature and things would still work the same. The idea of having an `abstractmethod` is just to enforce the concrete class to implement a method with the stated interface, the signature of that implementation applies to the concrete method, not to the abstractmethod. This is a normal case where `super` can be used to save the concrete implementations from having to implement stuff that is always needed and can be placed in a parent method.

Reply:
>  I find super to be more intuitive because it tells the user that the code in the parent method is useful for the code in the child method.

If the the abstractmethod *did* return a Counter instance, you would just be throwing it away, which doesn't make much sense. I agree it makes sense when the super class behavior composes with the child class's. That is not the case here. It's actually coupling the concrete class with the abstract class.

> This is a normal case where super can be used to save the concrete implementations from having to implement stuff that is always needed and can be placed in a parent method.

Agreed, I think a helper method which type checks and does what it claims to do is more intuitive. Just reading `super().create_counter(...)` I have no idea what the parent class impl does. If you want to force this behavior, a better way is to create a template method instead.

I'm not claiming it's not possible, just something we should avoid. Btw, mypy does complain about this kind of code because it violates the signature.

<img width="475" alt="Screen Shot 2022-03-09 at 11 31 59 AM" src="https://user-images.githubusercontent.com/1510004/157486348-900a8f1b-cf87-44cc-baa8-a3fe5bb1b994.png">

Reply:
Moved check invocation to SDK, please review again @aabmass